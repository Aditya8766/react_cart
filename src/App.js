import "./App.css";
import React, { useEffect, useState } from "react";
import Header from "./components/Header";
import Main from "./components/Main";
import Cart from "./components/Cart";
import products from "./mockdata/Products";
import Pagination from "./components/pagination";

//  function getcall(){
//   let url="https://dummyjson.com/products";
//   return fetch(url);
//  }
function App() {
  const [productData, setproductData] = useState([]);
  const [cartProducts, setcartProducts] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [postPerPage, setPostPerPage] = useState(3);

  function getProducts() {
    return new Promise((resolve) => {
      resolve(products);
    });
  }
  function processData(Promise) {
    Promise.then((data) => {
      setproductData(data.products);
    }).catch(Error);
  }
  useEffect(() => {
    processData(getProducts());
  }, []);

  const handleAddToCart = (product) => {
    const matched = cartProducts.find(
      (cartProduct) => cartProduct.id === product.id
    );
    if (matched) {
      matched.quantity = matched.quantity + 1;
      setcartProducts([...cartProducts]);
    } else {
      product.quantity = 1;
      setcartProducts([...cartProducts, product]);
    }
  };

  const handleAddToCart1 = (product) => {
    setcartProducts((prevCartProducts) => {
      const index = prevCartProducts.findIndex((p) => p.id === product.id);
      prevCartProducts.splice(index, 1);
      return [...prevCartProducts];
    });
  };

  const lastPostIndex = currentPage * postPerPage;
  const firstPostIndex = lastPostIndex - postPerPage;
  const currentPost = productData?.slice(firstPostIndex, lastPostIndex);

  return (
    <div className="App">
      <Header />
      <Main
        handleAddToCart={handleAddToCart}
        
        productData={currentPost}
      />
      <Pagination
        totalPosts={productData.length}
        postPerPage={postPerPage}
        setCurrentPage={setCurrentPage}
        currentPage={currentPage}
      />
      <Cart cartProducts={cartProducts}
       handleAddToCart1={handleAddToCart1} />
     
    </div>
  );
}
export default App;
