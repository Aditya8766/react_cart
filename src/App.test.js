import { render,screen,fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import App from './App';


describe("App",()=>{


  beforeEach(()=>{
    global.fetch = jest.fn(()=>{
      Promise.resolve({
        json: ()=>
        Promise.resolve({ products:[{id:"1",title:"abc",price:"530"}] })
      })
    });
  });
  afterAll(()=>{
    jest.clearAllMocks();
  });
  
  test('renders learn react link', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should add value when button is clicked",async()=>{   
   let app; 
   await act(async()=>{
    app = render(<App />);
  });
  const btn = app.container.querySelector("button")
  fireEvent.click(btn);
  
  });
  it("Should remove value when button is clicked",async()=>{   
    let app; 
    await act(async()=>{
     app = render(<App />);
   });
   const btn = app.container.querySelector(".product__btn2")
   fireEvent.click(btn);
   
   });
});

