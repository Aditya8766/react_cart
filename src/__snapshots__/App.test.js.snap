// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders learn react link 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div
      class="header"
    >
      <h1>
        Products
      </h1>
    </div>
    <div
      class="Main-block"
    />
    <div
      class="pagination"
    />
    <div
      class="cart"
    >
      <div
        class="total-cart-val"
      >
        <h3>
          Grand Total
        </h3>
        0
      </div>
    </div>
  </div>
</DocumentFragment>
`;
